lib/vsc/install/shared_setup.py

```diff
@@ -174,6 +187,123 @@ def run(self):
+            if mod in test_filter['allowmods']:
+                # a parent name space
+                pass
+            elif re.search(test_filter['module'], mod):
+                if test_filter['function'] is not None:
+                    res = filter_testsuites(testsuites)
+                # add parents (and module itself)
+                pms = mod.split('.')
+                for pm_idx in range(len(pms)):
+                    pm = '.'.join(pms[:pm_idx])
+                    if not pm in test_filter['allowmods']:
+                        test_filter['allowmods'].append(pm)
+            else:
+                res = type(testsuites)()
+
+        return res
+
+
+class VscTestCommand(TestCommand):
+    """
+    The cmdclass for testing
+    """
+    user_options = TestCommand.user_options + [
+        ('test-filterf=', 'f', "Regex filter on test function names"),
+        ('test-filterm=', 'F', "Regex filter on test (sub)modules"),
+    ]
+
+    def initialize_options(self):
+        TestCommand.initialize_options(self)
+        self.test_filterm = None
+        self.test_filterf = None
+        self.test_loader = "vsc.install.shared_setup:VscScanningLoader"
+
+    def run_tests(self):
+        getattr(__builtin__,'__test_filter').update({
+            'function': self.test_filterf,
+            'module': self.test_filterm,
+        })
+
+        # should be setup.py
+        setup_py = os.path.abspath(sys.argv[0])
+        log.info('run_tests from %s' % setup_py)
+        base_dir = os.path.dirname(setup_py)
+
+        # make a lib dir to trick setup.py to package this properly
+        # and git ignore empty dirs, so recreate it if necessary
+        lib_dir = os.path.join(base_dir, 'lib')
+        if not os.path.exists(lib_dir):
+            os.mkdir(lib_dir)
+
+        test_dir = os.path.join(base_dir, DEFAULT_TEST_SUITE)
+        if os.path.isdir(test_dir):
+            sys.path.insert(0, test_dir)
+        else:
+            raise Exception("Can't find location of testsuite directory %s in %s" % (DEFAULT_TEST_SUITE, base_dir))
+
+        # make sure we can import the script as a module
+        scripts_dir = os.path.join(base_dir, 'bin')
+        if os.path.isdir(scripts_dir):
+            sys.path.insert(0, scripts_dir)
+
+        # and now the ugly bits: cleanup and restore vsc namespace
+        #   we use vsc namespace tools very early
+        #   need to make sure they are picked up from the paths as specified above
+        #   not to mix with installed and already loaded modules
+        loaded_vsc_modules = [mod for mod in sys.modules.keys() if mod == 'vsc' or mod.startswith('vsc.')]
+        reload_vsc_modules = []
+        for mod in loaded_vsc_modules:
+            if hasattr(sys.modules[mod], '__file__'):
+                # only actual modules
+                reload_vsc_modules.append(mod)
+            del(sys.modules[mod])
+        # reimport
+        for mod in reload_vsc_modules:
+            __import__(mod)
```

Comment:
how do you know whether you're importing the right stuff here?

because of the sys.path hack at the top?

maybe check here whether things are being imported where you expect from (use `base_dir`)?